### PR TITLE
Add documentation for keyserver package

### PR DESCRIPTION
**Pull request by c633**

### CODE REVIEW

**Comment by arlolra** [comment]
Should we drop the README.md in this package?

**Comment by c633** [comment]
> Should we drop the README.md in this package?

I want to keep it here for people reading our doc on Github. Which do you prefer: reading doc on Github or godoc.com?

**Comment by arlolra** [comment]
> Which do you prefer: reading doc on Github or godoc.org?

I prefer not having to maintain two sets of documentation :P

But, so far, our godocs seems more complete, so I'd opt for that.

**Comment by arlolra** [comment]
But maybe my views are controversial.  Something else for the agenda tomorrow.

File: keyserver/encoding.go [hunk 1]
```diff
@@ -10,10 +10,13 @@ import (
 	. "github.com/coniks-sys/coniks-go/protocol"
 )
 
+// MarshalResponse returns the JSON encoding of the server's response.
 func MarshalResponse(response *Response) ([]byte, error) {
 	return json.Marshal(response)
 }
 
+// UnmarshalRequest parses the JSON-encoded request and
```
Comment (masomel) [inline keyserver/encoding.go]: I would change to "... *a* JSON-encoded client request..."

File: keyserver/encoding.go [hunk 2]
```diff
@@ -10,10 +10,13 @@ import (
...
+// creates the corresponding request which could be handled by the server.
```
Comment (masomel) [inline keyserver/encoding.go]: s/request/protocol.message.Request (I'm not sure if this is the right convention for citing interfaces in other packages). Also, s/could/will

File: keyserver/testutil/testutil.go [hunk 3]
```diff
@@ -1,3 +1,6 @@
+/*
+Package testutil provides utility functions for writing tests.
```
Comment (masomel) [inline keyserver/testutil/testutil.go]: What kinds of tests? Maybe clarify this.

File: keyserver/server.go [hunk 4]
```diff
@@ -21,21 +21,34 @@ import (
...
 type ServerConfig struct {
-	configFilePath      string
-	DatabasePath        string          `toml:"database"`
+	// DatabasePath is a path to the database for storing the server's directory.
+	DatabasePath string `toml:"database"`
+	// LoadedHistoryLength is the maximum length of current snapshots stored in memory.
 	LoadedHistoryLength uint64          `toml:"loaded_history_length"`
 	Policies            *ServerPolicies `toml:"policies"`
 	TLS                 *TLSConnection  `toml:"tls"`
+	configFilePath      string
```
Comment (masomel) [inline keyserver/server.go]: For completeness, I think you should add descriptions for `Policies` and `TLS`, too.

File: keyserver/server.go [hunk 5]
```diff
@@ -194,6 +221,7 @@ func (server *ConiksServer) EpochUpdate() {
 	}
 }
 
+// Shutdown closes all server's connections and shutdowns the server.
```
Comment (masomel) [inline keyserver/server.go]: s/all/all of the
s/shutdowns/shuts down

**Comment by masomel** [comment]
@c633 If you want to keep the information that was in merkletree/README.md, that could be moved to merkletree/doc.go?

**Comment by c633** [comment]
I thought we decided to remove merkletree/README.md?

File: keyserver/server.go (hunk 4 above)
Comment (c633) [inline keyserver/server.go]: I'm not sure if we want to have something just like "Policies is a pointer to ServerPolicies struct"?

**Comment by masomel** [comment]
> I thought we decided to remove merkletree/README.md?

Remove the README file, yes. My comment was more about if you wanted to keep the usage information for developers (which I think could be useful), the appropriate place would be merkletree/doc.go.

**Comment by c633** [comment]
Ahh no, that's my bad, sorry. I meant the current merkletree/README is mostly like an API docs and we decided to remove because of that.

File: keyserver/server.go (hunk 4 above)
Comment (masomel) [inline keyserver/server.go]: Maybe not quite like that, maybe something more like "Policies contains the server's CONIKS policies configuration"?

Comment (c633) [inline keyserver/server.go]: Sure! Thanks!

File: keyserver/server.go [hunk 6]
```diff
@@ -44,6 +59,13 @@ type ServerPolicies struct {
 	signKey       sign.PrivateKey
 }
 
+// A ConiksServer represents a CONIKS key server.
+// It listens for a public TCP connection which accepts
+// all operation-protocols except registration and a local
+// UNIX socket connection which accepts registration requests
+// from the account verification proxy.
```
Comment (arlolra) [inline keyserver/server.go]: It probably needs updating in light of #130

File: keyserver/server.go [hunk 7]
```diff
@@ -22,21 +22,42 @@ import (
...
+	Policies *ServerPolicies `toml:"policies"`
+	// Addresses contains the server's connections configuration.
+	Addresses      []*Address `toml:"addresses"`
+	configFilePath string
 }
 
+// An Address declares a server's connection.
+// It makes the server connections configurable
+// so that a key server implementation can easily
+// be used as a first-party identity provider or
```
Comment (masomel) [inline keyserver/server.go]: s/as/by?

File: keyserver/server.go [hunk 8]
```diff
@@ -22,21 +22,42 @@ import (
...
+// a third-party communication service.
+// It supports two types of connections: a TCP connection ("tcp")
+// and a Unix socket connection ("unix"). Allowing registration
+// has to be explicitly defined for each connection.
+// Other types of requests are allowed by default.
```
Comment (masomel) [inline keyserver/server.go]: I would add the explanation here that a registration is considered a "write", while the other requests are "reads", just to make it more clear why we treat registration specially.

File: keyserver/server.go [hunk 9]
```diff
@@ -45,6 +66,11 @@ type ServerPolicies struct {
 	signKey       sign.PrivateKey
 }
 
+// A ConiksServer represents a CONIKS key server.
+// It wraps a ConiksDirectory with a network layer which
+// handles requests/responses encoding/decoding. It also
+// supports handling concurrent requests and mechanism to
```
Comment (masomel) [inline keyserver/server.go]: s/mechanism/a mechanism

**Comment by c633** [comment]
Can I merge this one? We can fix the summary of `coniksserver` package in `fix-doc` later.

**Comment by masomel** [comment]
Yeah, this LGTM. Feel free to merge. Like you said, any other fixes can happen on the `fix-doc` branch.

**Comment by c633** [comment]
Merged in 0f88ea5171c55692dd72a98bf36b6891f34a61b3.

**Pull request closed without merging**